src/tuntap/iface.go

```diff
@@ -21,25 +21,50 @@ func (tun *TunAdapter) writer() error {
 		if n == 0 {
 			continue
 		}
-		if tun.iface.IsTAP() {
-			var dstAddr address.Address
-			if b[0]&0xf0 == 0x60 {
-				if len(b) < 40 {
-					//panic("Tried to send a packet shorter than an IPv6 header...")
-					util.PutBytes(b)
-					continue
-				}
-				copy(dstAddr[:16], b[24:])
-			} else if b[0]&0xf0 == 0x40 {
-				if len(b) < 20 {
-					//panic("Tried to send a packet shorter than an IPv4 header...")
-					util.PutBytes(b)
-					continue
-				}
-				copy(dstAddr[:4], b[16:])
-			} else {
-				return errors.New("Invalid address family")
+		var dstAddr address.Address
+		var addrlen int
+		// Check whether the packet is IPv4, IPv6 or neither
+		if b[0]&0xf0 == 0x60 {
+			// IPv6 packet found
+			if len(b) < 40 {
+				// Packet was too short
+				util.PutBytes(b)
+				continue
+			}
+			// Extract the destination IPv6 address
+			copy(dstAddr[:16], b[24:])
+			addrlen = 16
+		} else if b[0]&0xf0 == 0x40 {
+			// IPv4 packet found
+			if len(b) < 20 {
+				// Packet was too short
+				util.PutBytes(b)
+				continue
+			}
+			// Extract the destination IPv4 address
+			copy(dstAddr[:4], b[16:])
+			addrlen = 4
+		} else {
+			// Neither IPv4 nor IPv6
+			return errors.New("Invalid address family")
+		}
+		// Check the crypto-key routing rules next
+		if tun.ckr.isEnabled() {
+			if !tun.ckr.isValidLocalAddress(dstAddr, addrlen) {
+				util.PutBytes(b)
+				continue
+			}
```

Comment:
We should also check that the remote side's source address is one that we'll accept.

Reply:
I've added in a check here that the source address matches *one of the ranges* we have configured as a remote, but without having any context for which key gave us this TUN/TAP traffic I can't improve the check to verify the specific key. Therefore I'm not sure how much value the check really gives us.